plugins/thunderbolt/fu-thunderbolt-device.c

```diff
@@ -0,0 +1,645 @@
+					g_warning ("failed to get native mode status: %s",
+						   native_error->message);
+				}
+				domain_id = g_strdup_printf ("TBT-%04x%04x%s-controller%s",
+							     (guint) vid,
+							     (guint) did,
+							     is_native ? "-native" : "",
+							     domain);
+			}
+			vendor_id = g_strdup_printf ("TBT:0x%04X", (guint) vid);
+			fu_device_set_vendor_id (device, vendor_id);
+			device_id = g_strdup_printf ("TBT-%04x%04x%s",
+						     (guint) vid,
+						     (guint) did,
+						     is_native ? "-native" : "");
+			fu_device_add_flag (device, FWUPD_DEVICE_FLAG_UPDATABLE);
+			fu_device_add_flag (device, FWUPD_DEVICE_FLAG_DUAL_IMAGE);
+		} else {
+			device_id = g_strdup ("TBT-fixed");
+			fu_device_set_update_error (device, "Missing non-active nvmem");
+		}
+		fu_device_add_instance_id (device, device_id);
+		if (domain_id != NULL)
+			fu_device_add_instance_id (device, domain_id);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_plugin_thunderbolt_trigger_update (FuThunderboltDevice	*self,
+				      GError			**error)
+{
+
+	ssize_t n;
+	int fd;
+	int r;
+	g_autofree gchar *auth_path = NULL;
+
+	auth_path = g_build_filename (self->devpath, "nvm_authenticate", NULL);
+
+	fd = open (auth_path, O_WRONLY | O_CLOEXEC);
+	if (fd < 0) {
+		g_set_error (error, G_IO_ERROR,
+			     g_io_error_from_errno (errno),
+			     "could not open 'nvm_authenticate': %s",
+			     g_strerror (errno));
+		return FALSE;
+	}
+
+	do {
+		n = write (fd, "1", 1);
+		if (n < 1 && errno != EINTR) {
+			g_set_error (error, G_IO_ERROR,
+				     g_io_error_from_errno (errno),
+				     "could not write to 'nvm_authenticate': %s",
+				     g_strerror (errno));
+			(void) close (fd);
+			return FALSE;
+		}
+	} while (n < 1);
+
+	r = close (fd);
+	if (r < 0 && errno != EINTR) {
+		g_set_error (error, G_IO_ERROR,
+			     g_io_error_from_errno (errno),
+			     "could not close 'nvm_authenticate': %s",
+			     g_strerror (errno));
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_thunderbolt_device_attach (FuDevice *device, GError **error)
+{
+	const gchar *attribute;
+	guint64 status;
+
+	/* now check if the update actually worked */
+	attribute = fu_udev_device_get_sysfs_attr (FU_UDEV_DEVICE (device),
+						   "nvm_authenticate",
+						   error);
+	if (attribute == NULL)
+		return FALSE;
+	status = g_ascii_strtoull (attribute, NULL, 16);
+	if (status == G_MAXUINT64 && errno == ERANGE) {
+		g_set_error (error, G_IO_ERROR,
+			     g_io_error_from_errno (errno),
+			     "failed to read 'nvm_authenticate: %s",
+			     g_strerror (errno));
+		return FALSE;
+	}
+
+	/* anything else then 0x0 means we got an error */
+	if (status != 0x0) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "update failed (status %" G_GINT64_MODIFIER "x)", status);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_thunderbolt_device_write_data (FuThunderboltDevice	*self,
```

Comment:
This function I wonder if we should replace the code with FuChunks instead

Reply:
I don't think we can, it's not writing in blocks of a set size, it's just writing until write() returns and then advancing a counter with the number of bytes we wrote.